src/osemgrep/configuring/Semgrep_settings.ml

```diff
@@ -0,0 +1,97 @@
+type t = {
+  has_shown_metrics_notification : bool option;
+  api_token : string option;
+  anonymous_user_id : Uuidm.t;
+}
+
+let default_settings =
+  {
+    has_shown_metrics_notification = None;
+    api_token = None;
+    anonymous_user_id = Uuidm.v `V4;
+  }
+
+let _t = ref default_settings
+let ( let* ) = Result.bind
+
+let of_yaml = function
+  | `O data ->
+      let has_shown_metrics_notification =
+        List.assoc_opt "has_shown_metrics_notification" data
+      and api_token = List.assoc_opt "api_token" data
+      and anonymous_user_id = List.assoc_opt "anonymous_user_id" data in
+      let* has_shown_metrics_notification =
+        Option.fold ~none:(Ok None)
+          ~some:(function
+            | `Bool b -> Ok (Some b)
+            | _else -> Error (`Msg "has shown metrics notification not a bool"))
+          has_shown_metrics_notification
+      in
+      let* api_token =
+        Option.fold ~none:(Ok None)
+          ~some:(function
+            | `String s -> Ok (Some s)
+            | _else -> Error (`Msg "api token not a string"))
+          api_token
+      in
+      let* anonymous_user_id =
+        Option.fold
+          ~none:(Error (`Msg "no anonymous user id"))
+          ~some:(function
+            | `String s ->
+                Option.to_result ~none:(`Msg "not a valid UUID")
+                  (Uuidm.of_string s)
+            | _else -> Error (`Msg "anonymous user id is not a string"))
+          anonymous_user_id
+      in
+      Ok { has_shown_metrics_notification; api_token; anonymous_user_id }
+  | _else -> Error (`Msg "YAML not an object")
+
+let to_yaml { has_shown_metrics_notification; api_token; anonymous_user_id } =
+  `O
+    ((match has_shown_metrics_notification with
+     | None -> []
+     | Some v -> [ ("has_shown_metrics_notification", `Bool v) ])
+    @ (match api_token with
+      | None -> []
+      | Some v -> [ ("api_token", `String v) ])
+    @ [ ("anonymous_user_id", `String (Uuidm.to_string anonymous_user_id)) ])
+
+let settings = Fpath.v Semgrep_envvars.env.user_settings_file
+
+let get_contents () =
+  lazy
+    (_t :=
+       try
+         let _ = File.fullpath settings in
```

Comment:
what is the line for?

Reply:
checking whether the file exists, otherwise we get a ENOENT from the read_file.

Reply:
done in [6ba7cc1](https://github.com/returntocorp/semgrep/pull/7637/commits/6ba7cc17a9567ff72c190dea6a67d7074483904e)